File: tests/attestation.ts
/*
 * Copyright (c) 2025, Circle Internet Financial LTD All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { PublicKey } from "@solana/web3.js";
import * as anchor from "@coral-xyz/anchor";
import { SOLANA_DOMAIN } from "./constants";

// Attestation Set fixed header offsets
export const ATTESTATION_SET_MAGIC = 0x10cbb1ec;
export const MAGIC_OFFSET = 0;
export const VERSION_OFFSET = 4;
export const DESTINATION_DOMAIN_OFFSET = 8;
export const DESTINATION_CONTRACT_OFFSET = 12;
export const DESTINATION_CALLER_OFFSET = 44;
export const MAX_BLOCK_HEIGHT_OFFSET = 76;
export const ATTESTATION_SET_NUM_ATTESTATIONS_OFFSET = 84;
export const ATTESTATION_SET_ATTESTATIONS_OFFSET = 88;

// Attestation Element relative offsets
export const DESTINATION_TOKEN_OFFSET = 0;
export const DESTINATION_RECIPIENT_OFFSET = 32;
export const VALUE_OFFSET = 64;
export const TRANSFER_SPEC_HASH_OFFSET = 72;
export const HOOK_DATA_LENGTH_OFFSET = 104;
export const HOOK_DATA_OFFSET = 108;

export type MintAttestationSet = {
  version: number;
  destinationDomain: number;
  destinationContract: PublicKey;
  destinationCaller: PublicKey;
  maxBlockHeight: anchor.BN;
  numAttestations: number;
  attestations: MintAttestationElement[];
};

export type MintAttestationElement = {
  destinationToken: PublicKey;
  destinationRecipient: PublicKey;
  value: anchor.BN;
  transferSpecHash: Buffer;
  hookDataLength: number;
  hookData: Buffer;
};

export type MintAttestationSetParams = {
  isDefaultDestinationCaller: boolean;
  maxBlockHeight: number;
  elements: MintAttestationParams[];
};

export type MintAttestationParams = {
  value: number;
  transferSpecHash: Buffer;
  hookData: Buffer;
};

// Helper functions

/**
 * Generate a random bigint value (up to 64-bit safe integer)
 */
export function randomBigInt(): bigint {
  const randomValue = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
  return BigInt(randomValue);
}

/**
 * Convert an anchor.BN to an 8-byte big-endian buffer
 */
export function bnToBuffer8BE(value: anchor.BN): Buffer {
  const buffer = Buffer.alloc(8);
  const hex = value.toString(16, 16); // pad to 16 hex chars (8 bytes)
  Buffer.from(hex, "hex").copy(buffer);
  return buffer;
}

/**
 * Encode a MintAttestationElement to a buffer
 */
export function encodeMintAttestationElement(
  element: MintAttestationElement
): Buffer {
  const hook = Buffer.from(element.hookData);
  const len = HOOK_DATA_OFFSET + hook.length;
  const out = Buffer.alloc(len);

  element.destinationToken.toBuffer().copy(out, DESTINATION_TOKEN_OFFSET);
  element.destinationRecipient
    .toBuffer()
    .copy(out, DESTINATION_RECIPIENT_OFFSET);
  bnToBuffer8BE(element.value).copy(out, VALUE_OFFSET);
  Buffer.from(element.transferSpecHash).copy(out, TRANSFER_SPEC_HASH_OFFSET);
  out.writeUInt32BE(element.hookDataLength, HOOK_DATA_LENGTH_OFFSET);
  if (hook.length > 0) hook.copy(out, HOOK_DATA_OFFSET);
  return out;
}

/**
 * Encode a MintAttestationSet to a buffer
 */
export function encodeMintAttestationSet(set: MintAttestationSet): Buffer {
  const encodedAttestations = Buffer.concat(
    set.attestations.map(encodeMintAttestationElement)
  );
  const out = Buffer.alloc(
    ATTESTATION_SET_ATTESTATIONS_OFFSET + encodedAttestations.length
  );

  out.writeUInt32BE(ATTESTATION_SET_MAGIC, MAGIC_OFFSET);
  out.writeUInt32BE(set.version, VERSION_OFFSET);
  out.writeUInt32BE(set.destinationDomain, DESTINATION_DOMAIN_OFFSET);
  set.destinationContract.toBuffer().copy(out, DESTINATION_CONTRACT_OFFSET);
  set.destinationCaller.toBuffer().copy(out, DESTINATION_CALLER_OFFSET);
  bnToBuffer8BE(set.maxBlockHeight).copy(out, MAX_BLOCK_HEIGHT_OFFSET);
  out.writeUInt32BE(
    set.numAttestations,
    ATTESTATION_SET_NUM_ATTESTATIONS_OFFSET
  );
  encodedAttestations.copy(out, ATTESTATION_SET_ATTESTATIONS_OFFSET);
  return out;
}

/**
 * Decode a Buffer to a MintAttestationSet
 */
export function decodeMintAttestationSet(buffer: Buffer): MintAttestationSet {
  const version = buffer.readUInt32BE(VERSION_OFFSET);
  const destinationDomain = buffer.readUInt32BE(DESTINATION_DOMAIN_OFFSET);
  const destinationContract = new PublicKey(
    buffer.subarray(
      DESTINATION_CONTRACT_OFFSET,
      DESTINATION_CONTRACT_OFFSET + 32
    )
  );
  const destinationCaller = new PublicKey(
    buffer.subarray(DESTINATION_CALLER_OFFSET, DESTINATION_CALLER_OFFSET + 32)
  );
  const maxBlockHeight = new anchor.BN(
    buffer
      .subarray(MAX_BLOCK_HEIGHT_OFFSET, MAX_BLOCK_HEIGHT_OFFSET + 8)
      .toString("hex"),
    16
  );
  const numAttestations = buffer.readUInt32BE(
    ATTESTATION_SET_NUM_ATTESTATIONS_OFFSET
  );

  // Decode each attestation element
  const attestations: MintAttestationElement[] = [];
  let offset = ATTESTATION_SET_ATTESTATIONS_OFFSET;

  for (let i = 0; i < numAttestations; i++) {
    const destinationToken = new PublicKey(
      buffer.subarray(
        offset + DESTINATION_TOKEN_OFFSET,
        offset + DESTINATION_TOKEN_OFFSET + 32
      )
    );
    const destinationRecipient = new PublicKey(
      buffer.subarray(
        offset + DESTINATION_RECIPIENT_OFFSET,
        offset + DESTINATION_RECIPIENT_OFFSET + 32
      )
    );
    const value = new anchor.BN(
      buffer
        .subarray(offset + VALUE_OFFSET, offset + VALUE_OFFSET + 8)
        .toString("hex"),
      16
    );
    const transferSpecHash = buffer.subarray(
      offset + TRANSFER_SPEC_HASH_OFFSET,
      offset + TRANSFER_SPEC_HASH_OFFSET + 32
    );
    const hookDataLength = buffer.readUInt32BE(
      offset + HOOK_DATA_LENGTH_OFFSET
    );
    const hookData = buffer.subarray(
      offset + HOOK_DATA_OFFSET,
      offset + HOOK_DATA_OFFSET + hookDataLength
    );

    attestations.push({
      destinationToken,
      destinationRecipient,
      value,
      transferSpecHash,
      hookDataLength,
      hookData,
    });

    // Move to next attestation element
    offset += HOOK_DATA_OFFSET + hookDataLength;
  }

  return {
    version,
    destinationDomain,
    destinationContract,
    destinationCaller,
    maxBlockHeight,
    numAttestations,
    attestations,
  };
}

/**
 * Generate a random MintAttestationElement
 */
export function generateMintAttestationElement(
  overrides: Partial<MintAttestationElement> = {}
): MintAttestationElement {
  const randomBytes = (length: number) =>
    Buffer.from(Array.from({ length }, () => Math.floor(Math.random() * 256)));
  const randomPublicKey = () => new PublicKey(randomBytes(32));
  const hook = randomBytes(Math.floor(Math.random() * 33));

  const defaults: MintAttestationElement = {
    destinationToken: randomPublicKey(),
    destinationRecipient: randomPublicKey(),
    value: new anchor.BN(Math.floor(Math.random() * 100000000).toString()),
    transferSpecHash: randomBytes(32),
    hookDataLength: hook.length,
    hookData: hook,
  };
  return { ...defaults, ...overrides };
}

/**
 * Generate a random MintAttestationSet
 */
export function generateMintAttestationSet(
  overrides: Partial<MintAttestationSet> = {}
): MintAttestationSet {
  const randomBytes = (length: number) =>
    Buffer.from(Array.from({ length }, () => Math.floor(Math.random() * 256)));
  const randomPublicKey = () => new PublicKey(randomBytes(32));

  const attestations =
    overrides.attestations ||
    Array.from({ length: overrides.numAttestations || 1 }, () =>
      generateMintAttestationElement()
    );

  const defaults: MintAttestationSet = {
    version: 1,
    destinationDomain: SOLANA_DOMAIN,
    destinationContract: randomPublicKey(),
    destinationCaller: randomPublicKey(),
    maxBlockHeight: new anchor.BN(Number.MAX_SAFE_INTEGER),
    numAttestations: attestations.length,
    attestations: attestations,
  };
  return { ...defaults, ...overrides };
}

export function attestationSetToParams(
  set: MintAttestationSet
): MintAttestationSetParams {
  return {
    isDefaultDestinationCaller: set.destinationCaller.equals(PublicKey.default),
    maxBlockHeight: set.maxBlockHeight.toNumber(),
    elements: set.attestations.map((attestation) => ({
      value: attestation.value.toNumber(),
      transferSpecHash: attestation.transferSpecHash,
      hookData: attestation.hookData,
    })),
  };
}

File: programs/gateway-minter/src/attestation.rs
/*
 * Copyright (c) 2025, Circle Internet Financial LTD All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

//! MintAttestation
//!
//! This module implements encoding and decoding for ReducedMintAttestation messages.
//! All message encodings use **big-endian**.
//!
//! Constants:
//! - Set magic: `0x10cbb1ec` (bytes4(keccak256("circle.gateway.ReducedAttestationSet")))
//!
//! Attestation set layout:
//! ```
//! offset  size  field
//! 0       4     magic (0x10cbb1ec)
//! 4       4     version
//! 8       4     destination_domain
//! 12      32    destination_contract
//! 44      32    destination_caller
//! 76      8     max_block_height (u64)
//! 84      4     num_attestations
//! 88      ?     attestations (concatenated)
//! ```
//!
//! Attestation element layout:
//! ```
//! offset  size  field
//! 0       32    destination_token
//! 32      32    destination_recipient
//! 64      8     value (u64)
//! 72      32    transfer_spec_hash
//! 104     4     hook_data_length
//! 108     N     hook_data
//! ```

use crate::error::GatewayMinterError;
use anchor_lang::prelude::*;

#[derive(Clone, Debug)]
pub struct MintAttestation<'a> {
    data: &'a [u8],
    offset: usize,
    index: u32,
    num_elements: u32,
}

// Iterator
impl<'a> MintAttestation<'a> {
    pub const ATTESTATION_SET_MAGIC: u32 = 0x10cbb1ec;

    // Byte offsets of each field in the MintAttestationSet header
    const MAGIC_OFFSET: usize = 0;
    const VERSION_OFFSET: usize = 4;
    const DESTINATION_DOMAIN_OFFSET: usize = 8;
    const DESTINATION_CONTRACT_OFFSET: usize = 12;
    const DESTINATION_CALLER_OFFSET: usize = 44;
    const MAX_BLOCK_HEIGHT_OFFSET: usize = 76;
    const ATTESTATION_SET_NUM_ATTESTATIONS_OFFSET: usize = 84;
    const ATTESTATION_SET_ATTESTATIONS_OFFSET: usize = 88;

    // Relative byte offsets of each field in an attestation element
    const DESTINATION_TOKEN_OFFSET: usize = 0;
    const DESTINATION_RECIPIENT_OFFSET: usize = 32;
    const VALUE_OFFSET: usize = 64;
    const TRANSFER_SPEC_HASH_OFFSET: usize = 72;
    const HOOK_DATA_LENGTH_OFFSET: usize = 104;
    const HOOK_DATA_OFFSET: usize = 108;

    pub fn new(message_bytes: &'a [u8]) -> Result<Self> {
        // The smallest valid encoding is an attestation set with 1 attestation
        require_gte!(
            message_bytes.len(),
            Self::ATTESTATION_SET_ATTESTATIONS_OFFSET + Self::HOOK_DATA_OFFSET,
            GatewayMinterError::AttestationTooShort
        );

        let mut attestation = Self {
            data: message_bytes,
            offset: 0,
            index: 0,
            num_elements: 0,
        };

        require!(
            attestation.magic()? == Self::ATTESTATION_SET_MAGIC,
            GatewayMinterError::AttestationMagicMismatch
        );

        attestation.num_elements = attestation.num_attestations()?;
        attestation.offset = Self::ATTESTATION_SET_ATTESTATIONS_OFFSET;

        require_gt!(
            attestation.num_elements,
            0,
            GatewayMinterError::EmptyAttestationSet
        );

        Ok(attestation)
    }

    #[allow(clippy::should_implement_trait)]
    pub fn next(&mut self) -> Result<bool> {
        if self.index >= self.num_elements {
            return Ok(false);
        }

        // Advance to the next attestation based on size of the attestation element
        // Do not advance on the first call; just expose the first element
        if self.index > 0 {
            let hook_data_length = Self::u32_to_usize(self.hook_data_length()?)?;
            let attestation_length = Self::checked_add(Self::HOOK_DATA_OFFSET, hook_data_length)?;
            self.offset = Self::checked_add(self.offset, attestation_length)?;
        }
        self.index += 1;

        // Check that there are enough bytes to read the next attestation's fixed header
        let remaining_length = self.data.len() - self.offset;
        require_gte!(
            remaining_length,
            Self::HOOK_DATA_OFFSET,
            GatewayMinterError::AttestationTooShort
        );

        // Check that there are enough bytes to read the next attestation's hook data
        let hook_data_length = Self::u32_to_usize(self.hook_data_length()?)?;
        let attestation_length = Self::checked_add(Self::HOOK_DATA_OFFSET, hook_data_length)?;
        require_gte!(
            remaining_length,
            attestation_length,
            GatewayMinterError::AttestationTooShort
        );

        // If the next attestation is the last, check that it has no extraneous bytes
        if self.index == self.num_elements {
            require_eq!(
                self.offset + attestation_length,
                self.data.len(),
                GatewayMinterError::AttestationTooLong
            );
        }

        // Show the next attestation
        Ok(true)
    }
}

// Field accessors
impl<'a> MintAttestation<'a> {
    /// Returns the magic field for a MintAttestation or MintAttestationSet
    pub fn magic(&self) -> Result<u32> {
        self.read_u32(Self::MAGIC_OFFSET)
    }

    /// Returns the num_attestations field for a MintAttestationSet
    pub fn num_attestations(&self) -> Result<u32> {
        self.read_u32(Self::ATTESTATION_SET_NUM_ATTESTATIONS_OFFSET)
    }

    /// Returns version field
    pub fn version(&self) -> Result<u32> {
        self.read_u32(Self::VERSION_OFFSET)
    }

    /// Returns destination_domain field
    pub fn destination_domain(&self) -> Result<u32> {
        self.read_u32(Self::DESTINATION_DOMAIN_OFFSET)
    }

    /// Returns destination_contract field
    pub fn destination_contract(&self) -> Result<Pubkey> {
        self.read_pubkey(Self::DESTINATION_CONTRACT_OFFSET)
    }

    /// Returns destination_token field
    pub fn destination_token(&self) -> Result<Pubkey> {
        self.read_pubkey(Self::checked_add(
            self.offset,
            Self::DESTINATION_TOKEN_OFFSET,
        )?)
    }

    /// Returns destination_recipient field
    pub fn destination_recipient(&self) -> Result<Pubkey> {
        self.read_pubkey(Self::checked_add(
            self.offset,
            Self::DESTINATION_RECIPIENT_OFFSET,
        )?)
    }

    /// Returns destination_caller field
    pub fn destination_caller(&self) -> Result<Pubkey> {
        self.read_pubkey(Self::DESTINATION_CALLER_OFFSET)
    }

    /// Returns value field
    pub fn value(&self) -> Result<u64> {
        self.read_u64(Self::checked_add(self.offset, Self::VALUE_OFFSET)?)
    }

    /// Returns max_block_height field
    pub fn max_block_height(&self) -> Result<u64> {
        self.read_u64(Self::MAX_BLOCK_HEIGHT_OFFSET)
    }

    /// Returns transfer_spec_hash field
    pub fn transfer_spec_hash(&self) -> Result<[u8; 32]> {
        self.read_bytes::<32>(Self::checked_add(
            self.offset,
            Self::TRANSFER_SPEC_HASH_OFFSET,
        )?)
    }

    /// Returns hook_data_length field
    pub fn hook_data_length(&self) -> Result<u32> {
        self.read_u32(Self::checked_add(
            self.offset,
            Self::HOOK_DATA_LENGTH_OFFSET,
        )?)
    }

    /// Returns hook_data field
    pub fn hook_data(&self) -> Result<&[u8]> {
        let hook_data_offset = Self::checked_add(self.offset, Self::HOOK_DATA_OFFSET)?;
        let hook_data_length = Self::u32_to_usize(self.hook_data_length()?)?;
        Ok(&self.data[hook_data_offset..Self::checked_add(hook_data_offset, hook_data_length)?])
    }

    // Private helpers

    /// Reads u32 field at the given offset
    fn read_u32(&self, index: usize) -> Result<u32> {
        let end = Self::checked_add(index, 4)?;
        Ok(u32::from_be_bytes(
            self.data[index..end]
                .try_into()
                .map_err(|_| error!(GatewayMinterError::MalformedMintAttestation))?,
        ))
    }

    /// Reads u64 field at the given offset
    fn read_u64(&self, index: usize) -> Result<u64> {
        let end = Self::checked_add(index, 8)?;
        Ok(u64::from_be_bytes(
            self.data[index..end]
                .try_into()
                .map_err(|_| error!(GatewayMinterError::MalformedMintAttestation))?,
        ))
    }

    /// Reads pubkey field at the given offset
    fn read_pubkey(&self, index: usize) -> Result<Pubkey> {
        Pubkey::try_from(
            &self.data[index..Self::checked_add(index, std::mem::size_of::<Pubkey>())?],
        )
        .map_err(|_| error!(GatewayMinterError::MalformedMintAttestation))
    }

    /// Reads bytes field at the given offset
    fn read_bytes<const N: usize>(&self, index: usize) -> Result<[u8; N]> {
        self.data[index..Self::checked_add(index, N)?]
            .try_into()
            .map_err(|_| error!(GatewayMinterError::MalformedMintAttestation))
    }

    /// Converts a u32 to a usize
    fn u32_to_usize(value: u32) -> Result<usize> {
        usize::try_from(value).map_err(|_| error!(GatewayMinterError::MalformedMintAttestation))
    }

    #[inline]
    fn checked_add(a: usize, b: usize) -> Result<usize> {
        a.checked_add(b)
            .ok_or_else(|| error!(GatewayMinterError::MalformedMintAttestation))
    }
}

#[derive(Clone, Debug)]
pub struct MintAttestationStruct<'a> {
    pub version: u32,
    pub destination_domain: u32,
    pub destination_contract: [u8; 32],
    pub destination_caller: [u8; 32],
    pub max_block_height: u64,
    pub elements: Vec<MintAttestationElementStruct<'a>>,
}

#[derive(Clone, Debug)]
pub struct MintAttestationElementStruct<'a> {
    pub destination_token: [u8; 32],
    pub destination_recipient: [u8; 32],
    pub value: u64,
    pub transfer_spec_hash: [u8; 32],
    pub hook_data: &'a [u8],
}

#[allow(clippy::too_many_arguments)]
impl<'a> MintAttestationStruct<'a> {
    pub fn encode_attestation(&self) -> Vec<u8> {
        let num_elements = self.elements.len() as u32;

        let mut total_size = MintAttestation::ATTESTATION_SET_ATTESTATIONS_OFFSET; // fixed header size
        for element in &self.elements {
            let hook_data_length = element.hook_data.len();
            total_size += MintAttestation::HOOK_DATA_OFFSET + hook_data_length;
        }

        let mut buffer = Vec::with_capacity(total_size);

        // Encode attestation set header
        buffer.extend_from_slice(&MintAttestation::ATTESTATION_SET_MAGIC.to_be_bytes());
        buffer.extend_from_slice(&self.version.to_be_bytes());
        buffer.extend_from_slice(&self.destination_domain.to_be_bytes());
        buffer.extend_from_slice(&self.destination_contract);
        buffer.extend_from_slice(&self.destination_caller);
        buffer.extend_from_slice(&self.max_block_height.to_be_bytes());
        buffer.extend_from_slice(&num_elements.to_be_bytes());

        // Encode each attestation element
        for element in &self.elements {
            let hook_data_length = element.hook_data.len() as u32;
            buffer.extend_from_slice(&element.destination_token);
            buffer.extend_from_slice(&element.destination_recipient);
            buffer.extend_from_slice(&element.value.to_be_bytes());
            buffer.extend_from_slice(&element.transfer_spec_hash);
            buffer.extend_from_slice(&hook_data_length.to_be_bytes());
            buffer.extend_from_slice(element.hook_data);
        }

        buffer
    }
}
